### ISSUE TITLE
Poor compression & quality for difficult-to-compress data

### ISSUE THREAD

**Issue by lindstro** [CODARcode/MGARD#189]
I am doing some compression studies that involve difficult-to-compress (even incompressible) data.  Consider the chaotic data generated by the logistic map _x_<sub>_i_+1</sub> = 4 _x_<sub>_i_</sub> (1 - _x_<sub>_i_</sub>):
```
#include <cstdio>

int main()
{
  double x = 1. / 3;
  for (int i = 0; i < 256 * 256 * 256; i++) {
    fwrite(&x, sizeof(x), 1, stdout);
    x = 4 * x * (1 - x);
  }
  return 0;
}
```
We wouldn't expect this data to compress at all, but the inherent randomness at least suggests a predictable relationship between _L_<sub>2</sub> error, _E_, and rate, _R_.  Let &sigma; = 1/&Sqrt;8 denote the standard deviation of the input data and define the _accuracy gain_ as

&alpha; = log&#x2082;(&sigma; / _E_) - _R_.

Then each increment in storage, _R_, by one bit should result in a halving of _E_, so that &alpha; is essentially constant.  The limit behavior is slightly different as _R_ &rarr; 0 or _E_ &rarr; 0, but over a large range &alpha; ought to be constant.

Below is a plot of &alpha;(_R_) for MGARD 1.2.0 and other compressors applied to the above data interpreted as a 3D array of size 256 &times; 256 &times; 256.  Here I used a smoothness parameter of 0, which should result in an _L_<sub>2</sub> optimal reconstruction: `mgard compress --smoothness 0 --tolerance tolerance --datatype double --shape 256x256x256 --input input.bin --output output.mgard`.  The _tolerance_ was halved for each subsequent data point, starting with _tolerance_ = 1.

The plot suggests an odd relationship between _R_ and &alpha;, where &alpha; is far from stable when _R_ > 17.  Is this perhaps a bug in MGARD?  Similar behavior is observed for other difficult-to-compress data sets (see https://github.com/rballester/tthresh/issues/7).

![logistic](https://user-images.githubusercontent.com/1304791/160167361-4aea185e-9053-4ac6-ade6-f3cceda071fe.png)

**Comment by lindstro** [comment]
Just a friendly reminder to look into this issue when you get a chance.

**Comment by ben-e-whitney** [comment]
Apologies for the delay. I have begun working on this and have reproduced the problem.

![image](https://user-images.githubusercontent.com/5350974/165541830-ecfefec0-dd61-4a32-b4ba-6b42349f6ac5.png)

I'm guessing there is some issue with the coefficient quantization. Looking into it now.

In the meantime, I have a question about accuracy gain.

> … the inherent randomness at least suggests a predictable relationship between L2 error, E, and rate … each increment in storage, R, by one bit should result in a halving of E, so that α is essentially constant. The limit behavior is slightly different as R → 0 or E → 0, but over a large range α ought to be constant.

Under what circumstances "should" `α(R)` be about constant with respect to `R`? Is that the best possible behavior, or will it be achieved by a compressor well-suited to chaotic data or something? I am unfamiliar with accuracy gain and am trying to determine the point at which I should be pointing the finger at the algorithm rather than its implementation.

**Comment by lindstro** [comment]
> Under what circumstances "should" `α(R)` be about constant with respect to `R`? Is that the best possible behavior, or will it be achieved by a compressor well-suited to chaotic data or something? I am unfamiliar with accuracy gain and am trying to determine the point at which I should be pointing the finger at the algorithm rather than its implementation.

When _R_ = 0, no compressed data is stored, so the best you can hope for is that the compression algorithm (with no knowledge about the input data) by luck reconstructs the decompressed data as a constant _x_ = &mu; for all _x_, where &mu; is the mean of the input data.  In this case, _E_ = &sigma;, so &alpha; = 0.  Any other constant _x_ will yield &alpha; < 0.

For large _R_, either _E_ = 0 or _E_ converges to some small positive value, e.g., due to roundoff error in the compression scheme.  Hence, &alpha; either approaches +&infin; (_E_ = 0) or -&infin; (_E_ > 0) as _R_ grows. 

In between these limit behaviors, we would expect &alpha; to initially rise when _R_ is small, indicating that the data is effectively being compressed, until it reaches a plateau, where &alpha; is constant.  On this plateau, each increase in _R_ by one bit/value results in a halving of _E_, so &alpha; remains constant.  The value of this plateau is the number of per-value bits of information that the compressor has inferred and that need not be coded; it is a measure of how well the data has been compressed (higher is better).  Each additional bit coded is at this point random and so cannot be compressed (one bit of output stored for each bit of input consumed).  Eventually, &alpha; either goes to infinity or drops linearly.

The plot below illustrates this behavior for some compressors (in particular, SPERR and zfp) using the rather smooth Miranda viscosity data from [SDRBench](https://sdrbench.github.io), while other compressors do not reach a stable plateau (e.g., MGARD, SZ, TTHRESH).  Perhaps MGARD would if it weren't for this apparent bug.  We have diagnosed the TTHRESH bug and are working on a fix.

Of course, for random data, we do not expect any bits to be inferred, so the plateau usually is at or below zero.  Below zero because of overhead (e.g., metadata) and the pigeonhole principle; some outputs must be larger than the inputs in any compression scheme.  That's what we're seeing for all compressors in the plot above.

Think of accuracy gain as simply "rotating" (or, more accurately, "shearing") a PSNR vs. _R_ plot so that the expected 6.02 dB/bit slope turns into a slope of zero.

![viscosity](https://user-images.githubusercontent.com/1304791/165660745-d3ad7564-9dbe-4df4-96be-cd27ba1c6a40.png)

**Comment by ben-e-whitney** [comment]
Hi @lindstro. Apologies again for how long this is taking. Fixing the Huffman code ended up entailing a *lot* of work; see #196. Running on that branch, the accuracy gain plot looks a bit better.

![image](https://user-images.githubusercontent.com/5350974/177372621-8d74517a-bd7f-4dbe-aedd-c01e029f6897.png)

Comparison with what you were seeing:

* The initial plateau now extends to around 20 bits/double. In your tests, it only extended to around 17 bits/double.
* `α` falls after the initial plateau and then flattens out again around 28 bits/value, in keeping with your tests. In your tests, `α` then cratered and MGARD wasn't able to achieve compression ratios below 2. Now, we have a second 'plateau' (not that flat) lasting until about 58 bits/double, beyond which MGARD's quantization routine fails because its output can't be stored in the designated integer type.
* In the second plateau there's some strange degradation of `α` as the rate increases. It looks like the error tolerance decreases slightly and the compression ratio falls and/or the achieved error jumps. (I'm making the data for these plots by running MGARD with a sequence of error *tolerances,* while `α` is defined in terms of the compression ratio and the *achieved* error.) Then the tolerance decreases a bit more and things return to normal. I don't know what's going on here. 

Here's a summary of how the compression happens:

* After the decorrelation stage, the transform coefficients are quantized, yielding a sequence of integers.
* A Huffman code is built for the integer sequence. A certain range of integers is eligible for a Huffman codeword. In the configuration used to produce the plot above, that range is `{−2^17, …, 2^17 − 1}`. Integers inside this range are represented in a 'hit' buffer by the corresponding codeword. Integers outside this range are instead stored verbatim in a 'missed' buffer.
* The 'hit' buffer, the 'missed' buffer, and a bunch of metadata are written to a serialization buffer. This serialization buffer is then compressed by ZSTD.

As the error tolerance decreases, the quantized coefficients become larger, and as a result more of them fall outside the range `{−2^17, …, 2^17 − 1}` and end up in the 'missed' buffer. I guessed that the first plateau might end when this happened to some critical fraction of the quantized coefficients, so I repeated the experiment with some different integer ranges.

![image](https://user-images.githubusercontent.com/5350974/177372643-b4a8f5e4-f3f2-4b39-a96c-28e14c77528a.png)

I think we can say that this range parameter has some effect on the accuracy gain, but it's not the whole story.

How's all this looking to you?

**Comment by lindstro** [comment]
Thanks, @ben-e-whitney, for all your work in addressing this issue.  The accuracy gain plot looks a bit odd, but at least it doesn't go off a cliff, so that's a welcome improvement.  Let me play with this a bit and see how MGARD does on a few data sets.  Unfortunately, I cannot get the `huffman-decoding` branch to build on my Mac:
```
MGARD/include/huffman.tpp:183:14: fatal error: no viable conversion from 'std::shared_ptr<CodeCreationTreeNode>' to 'const bool'
  const bool children = node->left;
             ^          ~~~~~~~~~~
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.1.sdk/usr/include/c++/v1/memory:2879:22: note: explicit conversion function is not a candidate
    _LIBCPP_EXPLICIT operator bool() const _NOEXCEPT {return get() != nullptr;}
```

**Comment by ben-e-whitney** [comment]
It looks like my compiler is letting me get away with something a little improper. Once the branch is ready to be merged I'll have someone with a Mac try building and iron out any issues that come up. In the meantime, I think it'll compile with `static_cast<bool>(node->left)` in place of `node->left`.

**Comment by lindstro** [comment]
Thanks, but this leads to another issue:
```
MGARD/src/lossless_dispatcher.cpp:69:3: fatal error: static_assert failed due to requirement 'std::is_same<long long, long>::value' "deprecated Huffman coding written with assumption that `std::int64_t` is `long int`"
  static_assert(std::is_same<std::int64_t, long int>::value,
  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
Would be great if you could ensure MGARD builds on Mac.  I've heard from others that there might be linker issues as well.

**Comment by lindstro** [comment]
Here's an update.  I was able to get around the `int64_t` issue just by commenting the assertion out, which should be safe given that these types have the same size on my platform (Apple clang).  The MGARD `huffman-decoding` branch now builds, albeit slowly (took 36 minutes!).

I'm including two plots below using the SDRBench Miranda viscosity field and this command line:
```
mgard-x -z -s 0 -m abs -e TOLERANCE -t d -n 3 256 384 384 -r 0 -l 3 -d serial -i viscosity.raw -c viscosity.mgard
```
where TOLERANCE = 2.62133 / 2<sup>_i_</sup> for nonnegative integers _i_.  The first plot shows rate vs. accuracy gain.  MGARD goes off the cliff, as before, but at a lower rate this time.  The second plot shows that the absolute error tolerance is almost always exceeded.

By the way, the `-r` option is required but not documented.  I guessed that a value of zero would be a good default.
![viscosity](https://user-images.githubusercontent.com/1304791/179630116-146cb59b-347e-4e83-a575-c803d976df7c.png)
![mgard-error](https://user-images.githubusercontent.com/1304791/179630130-47e2aa87-198c-43ad-adb8-54c27a1fd323.png)

**Comment by lindstro** [comment]
@ben-e-whitney It's been some time since I saw progress on this issue or #186.  Do you have any updates to share?

**Comment by lindstro** [comment]
@ben-e-whitney Any progress on this issue?

**Comment by lindstro** [comment]
Can someone on the MGARD team please comment on the status of this issue?  It appears unresolved in the latest 1.3 release.
